### PR TITLE
Add logic that cleans the virtual services from cloud director 

### PR DESCRIPTION
**Pull request by jkremser**
It deletes the virtual services belonging to the cluster that is about to be deleted

Signed-off-by: Jiri Kremser <jiri.kremser@gmail.com>

## Checklist

~- [ ] Update changelog in CHANGELOG.md.~ (no releases yet)

### CODE REVIEW

File: pkg/cleaner/virtualservices.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		return true, microerror.Mask(fmt.Errorf(".status.infraId is not populated on the cluster: %s", c.Name))
+	}
+	deleted := 0
+	for _, vSvc := range vSvcs {
+		svcName := vSvc.NsxtAlbVirtualService.Name
+		//todo: pools
+		// if the name of the virtual service contains the cluster's infraId, delete this virtual service
+		if strings.Contains(svcName, infraId) {
+			log.Info(fmt.Sprintf("deleting virtual service: %s", svcName))
+			err = gateway.DeleteVirtualService(ctx, svcName, false)
```
Comment (erkanerol) [inline pkg/cleaner/virtualservices.go:47]: Does it work in an async way? Can we assume the LB is deleted when this call returns? If so, can we return true if all LBs are deleted correctly without waiting for next reconciliation loop?

File: pkg/vcd/vcd_utils.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"github.com/vmware/cloud-provider-for-cloud-director/pkg/vcdsdk"
+	capvcd "github.com/vmware/cluster-api-provider-cloud-director/api/v1beta1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func getUserCredentialsForCluster(ctx context.Context, cli client.Client, definedCreds capvcd.UserCredentialsContext) (capvcd.UserCredentialsContext, error) {
```
Comment (erkanerol) [inline pkg/vcd/vcd_utils.go:18]: Is this copy-paste from capvcd?

Comment (erkanerol) [reply pkg/vcd/vcd_utils.go:18]: Isn't there a function in the SDK we can call?

**Review by erkanerol — COMMENTED** [review]
The logic seems OK to me.

File: pkg/cleaner/virtualservices.go (hunk 1 above)
Comment (jkremser) [reply pkg/cleaner/virtualservices.go:47]: the false param here means "don't fail if it's not there", but the whole clean function returns fals in the "happy path scenario" so it's not being re-queued.

y, I think these calls are async, but will double check. Yesterday, I had some troubles to create a healthy vcdcluster on glados so it was hard to check from debugger, but the vServices got cleaned

**Comment by vxav** [comment]
Not sure if it's planned but I think we should probably make sure we delete the PVCs before the LB ?

**Comment by erkanerol** [comment]
> Not sure if it's planned but I think we should probably make sure we delete the PVCs before the LB ?

How are they related?

**Comment by vxav** [comment]
If we kill the LB we no longer have access to the cluster to remove the disks.

**Comment by erkanerol** [comment]
We are not deleting any disk in this operator now. CAPVCD doesn't delete them via WC API too. CAPVCD deletes them from cloud-director API directly.  I didn't get the point.

**Comment by jkremser** [comment]
the disks' leftovers in cloud director should be [deleted first](https://github.com/giantswarm/cluster-api-cleaner-cloud-director/blob/master/main.go#L101), the calls may be async so there can be non-deterministic order what happens first (I haven't checked). However, I don't think that deleting virtual services may be any issue and would make the cluster inaccessible. We are not cleaning the LB per se in the PR, but virtual services that are created if user creates a service that is "LBed". The LB itself will be untouched by this.

Cleaner for disks will be done in another PR (+ the order)🖖

**Comment by erkanerol** [comment]
@jkremser According to our tests, we don't need to delete volumes. They are deleted by capvcd manager somehow. Isn't it the case?

**Comment by jkremser** [comment]
> They are deleted by capvcd manager somehow. Isn't it the case?

@erkanerol it was broken on Wed so no idea, the PVs for PVCs were not able bound. But when you were showing me the issue, I recall you talking about some leftovers from PVs. But hey, this PR is not about volumes at all, so can I get the review on this one? ;)

**Review by erkanerol — APPROVED** [review]

**Pull request merged**